cpp/src/arrow/dataset/dataset_internal.h

```diff
@@ -52,53 +53,59 @@ class RecordBatchProjector {
  public:
   /// A column required by the given schema but absent from a record batch will be added
   /// to the projected record batch with all its slots null.
-  RecordBatchProjector(MemoryPool* pool, std::shared_ptr<Schema> to)
-      : pool_(pool),
-        to_(std::move(to)),
+  explicit RecordBatchProjector(std::shared_ptr<Schema> to)
+      : to_(std::move(to)),
         missing_columns_(to_->num_fields(), nullptr),
         column_indices_(to_->num_fields(), kNoMatch),
         scalars_(to_->num_fields(), nullptr) {}
 
-  /// A column required by the given schema but absent from a record batch will be added
-  /// to the projected record batch with all its slots equal to the corresponding entry in
-  /// scalars or null if the correspondign entry in scalars is nullptr.
-  RecordBatchProjector(MemoryPool* pool, std::shared_ptr<Schema> to,
-                       std::vector<std::shared_ptr<Scalar>> scalars)
-      : pool_(pool),
-        to_(std::move(to)),
-        missing_columns_(to_->num_fields(), nullptr),
-        column_indices_(to_->num_fields(), kNoMatch),
-        scalars_(std::move(scalars)) {
-    DCHECK_EQ(scalars_.size(), missing_columns_.size());
+  /// If the named field is absent from a record batch it will be added to the projected
```

Comment:
I don't agree with this premise. If it ain't present, you go the wrong `to` schema.

Reply:
You're saying that if the `to` schema has a field "alpha" but a record batch to be projected does not include a corresponding column then that should be an error?